lib/Dialect/TritonGPU/Transforms/LoopScheduling.cpp

```diff
@@ -0,0 +1,520 @@
+      // assumption is held by how loadOpsToIndirectionLevelAndUse recursively
+      // collects loadOpToIndLevelAndUse using DFS.
+      if (loadsToPipeline.count(loadOp) == 0) {
+        continue;
+      }
+    }
+
+    // If we still don't have a shared encoding, try a "generic" shared
+    // encoding.
+    if (!hasSharedEncoding && !isa<ttng::WarpGroupDotOp>(use)) {
+      hasSharedEncoding = hasSharedEncodingHelper(op, /*isMMAV3=*/loadIsMMAV3);
+    }
+
+    // If that still didn't work, bail on pipelining this load.
+    if (!hasSharedEncoding) {
+      continue;
+    }
+    loadsToPipeline.insert(op);
+  }
+  return loadsToPipeline;
+}
+
+static void scheduleLoads(scf::ForOp forOp, tt::CoarseSchedule &schedule,
+                          DenseSet<Operation *> &rootUsers, int numStages) {
+  ModuleOp moduleOp = forOp->getParentOfType<ModuleOp>();
+  tt::ModuleAxisInfoAnalysis axisInfoAnalysis(moduleOp);
+
+  // Get all loads that are (transitively) used by dot ops and their distance
+  // to the dot op.
+  llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
+      loadOpToIndLevelAndUse =
+          mlir::triton::loadOpsToIndirectionLevelAndUse(forOp);
+  LLVM_DEBUG({
+    LDBG("Found " << loadOpToIndLevelAndUse.size() << " loads to pipeline:");
+    for (const auto &[l, i, u] : loadOpToIndLevelAndUse) {
+      LDBG("  - load: " << *l);
+      LDBG("    at indirection level: " << i);
+      LDBG("    used by op: " << *u);
+    }
+  });
+  if (loadOpToIndLevelAndUse.empty())
+    return;
+
+  llvm::DenseSet<Operation *> loadsToPipeline =
+      filterPipelinedLoad(loadOpToIndLevelAndUse, axisInfoAnalysis);
+  if (loadsToPipeline.empty())
+    return;
+
+  // Calculate the stage distance between applicable loads.
+  int maxIndirectionLevel = -1;
+  for (auto [loadOp, dist, use] : loadOpToIndLevelAndUse) {
+    if (loadsToPipeline.count(loadOp) == 0)
+      continue;
+    maxIndirectionLevel = std::max(maxIndirectionLevel, dist);
+  }
+  unsigned stagesBetweenLoads =
+      ceil<unsigned>(numStages - 2, maxIndirectionLevel + 1);
+
+  tt::CoarseSchedule::Cluster rootUsersCluster = schedule.clusters.newAtFront();
+  // Put the root uses of the loads in the last stage.
+  for (auto &[loadOp, dist, use] : loadOpToIndLevelAndUse) {
+    if (loadsToPipeline.count(loadOp) == 0)
+      continue;
+    // Non-LoadOp(s) are the root uses of all LoadOp(s) and should be
+    // always present in the opInfo
+    if (!isa<tt::LoadOp>(use)) {
+      rootUsers.insert(use);
+      schedule.insert(use, numStages - 1, rootUsersCluster);
+    }
+  }
+
+  SmallVector<tt::CoarseSchedule::Cluster> loadsClusters;
+  for (int i = 0; i < maxIndirectionLevel + 1; i++) {
+    loadsClusters.push_back(schedule.clusters.newAtBack());
+  }
+  // Assign stages to the loads.
+  auto ctx = forOp.getContext();
+  for (auto [loadOp, indLevel, _] : loadOpToIndLevelAndUse) {
+    if (loadsToPipeline.count(loadOp) == 0)
+      continue;
+    loadOp->setAttr("loop.load",
```

Comment:
Should we have these strings defined in some shared file?

Reply:
What is the meaning of loop.load? Ideally we should only annotate with stage and schedule information otherwise we are adding ad hoc conventions between the different passes.